consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java

```diff
@@ -14,11 +14,135 @@
+import org.hyperledger.besu.ethereum.ProtocolContext;
+import org.hyperledger.besu.ethereum.core.Address;
+import org.hyperledger.besu.ethereum.core.Block;
+import org.hyperledger.besu.ethereum.mainnet.HeaderValidationMode;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
 public class RoundChangeMessageValidator {
 
-  public boolean validateRoundChange(final RoundChange msg) {
+  private static final String ERROR_PREFIX = "Invalid RoundChange Message";
+
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final RoundChangePayloadValidator roundChangePayloadValidator;
+  private final long quorumMessageCount;
+  private final long chainHeight;
+  private final Collection<Address> validators;
+  private final BlockValidator blockValidator;
+  private final ProtocolContext protocolContext;
+
+  public RoundChangeMessageValidator(
+      final RoundChangePayloadValidator roundChangePayloadValidator,
+      final long quorumMessageCount,
+      final long chainHeight,
+      final Collection<Address> validators,
+      final BlockValidator blockValidator,
+      final ProtocolContext protocolContext) {
+    this.roundChangePayloadValidator = roundChangePayloadValidator;
+    this.quorumMessageCount = quorumMessageCount;
+    this.chainHeight = chainHeight;
+    this.validators = validators;
+    this.blockValidator = blockValidator;
+    this.protocolContext = protocolContext;
+  }
+
+  public boolean validate(final RoundChange msg) {
+
+    if (!roundChangePayloadValidator.validate(msg.getSignedPayload())) {
+      LOG.info("{}: embedded payload was invalid", ERROR_PREFIX);
+      return false;
+    }
+
+    if (msg.getProposedBlock().isPresent()) {
+      return validateWithBlock(msg);
+    }
+
+    return msg.getPreparedRoundMetadata().isEmpty();
+  }
+
+  private boolean validateBlock(final Block block) {
+    final Optional<BlockProcessingOutputs> validationResult =
+        blockValidator.validateAndProcessBlock(
+            protocolContext, block, HeaderValidationMode.LIGHT, HeaderValidationMode.FULL);
+
+    if (validationResult.isEmpty()) {
+      LOG.info("{}: block did not pass validation.", ERROR_PREFIX);
+      return false;
+    }
+
     return true;
   }
+
+  private boolean validateWithBlock(final RoundChange msg) {
+    final Block block = msg.getProposedBlock().get();
+
+    if (!validateBlock(block)) {
+      return false;
+    }
+
+    if (msg.getPreparedRoundMetadata().isEmpty()) {
+      LOG.info("{}: Prepared block specified, but prepared metadata absent", ERROR_PREFIX);
+      return false;
+    }
+
+    final PreparedRoundMetadata metadata = msg.getPreparedRoundMetadata().get();
+
+    if (!metadata.getPreparedBlockHash().equals(block.getHash())) {
+      LOG.info("Invalid RoundChange message: Prepared metadata hash does not match supplied block");
```

Comment:
nit: Use ERROR_PREFIX as utilized in other logging messages.

Reply:
done